### DISCUSSION TITLE
[Proposal]: Partial-block only accessibility (private blockonly)

### DISCUSSION THREAD

**Discussion by TactiTac0z** [dotnet/csharplang#6394] [Language Ideas]
**Summary**
A new accessibility keyword for members (or types) used within partial types similar to `private` except it limits accessibility to the "partial-block" in which it was defined.

Consider the example:
```cs
public partial class Test 
{
  private int x;
  void ValidTestMethod()
  {
    x++;
  }
}

//Somewhere else in our file or in a different file:
partial class Test
{
  void BrokenTestMethod()
  {
    x++;
  }
}
```
So far this would still be allowed, but image we replaced `private` with a new keyword (lets call it `private blockonly` for now, I'll gladly take suggestions for what the best name for the keyword would be) such that our `x` field looks like this:
```cs
private blockonly int x;
```

The increment (the reference to `x`) in our `BrokenTestMethod` method would generate an accessibility compiler error because `x` is defined outside of the current "partial-block"
But our `ValidTestMethod` won't because `x` is defined within the same "partial-block".

The same applies to any member defined in a partial type (including types)
A class for instance would just simply be defined as follows:
```cs
partial class Test 
{
  private blockonly class Test2 {}
}
```

The existing behaviour of `private` (both in partial types and non-partial types) will of course remain the same

Our new keyword would of course be stackable with other keyword if the user so desires such as:
```cs
private blockonly readonly ref struct TestStruct {}
```
just to name a long example.
Although it would **NOT** be stackable with other accessibility keywords like `internal`, `protected`, etc.

**Motivation**
I personally use partial classes a lot, having a part of my type in multiple files (or perhaps multiple places within 1 file) is my preferred way to organize my code. Some developers like making "base classes" or perhaps "static helpers classes" of which I do a fair bit too, but I could really get a lot of use of this new feature even while mixing some of those techniques in.

Like for instances if I write a method that needs fields, but those fields only gets accessed within that method or within that area of my code, blockonly would help to prevent polluting the rest of my type when writing features or methods within that type.
Another Way is to append such fields with `_` or even `__` which is generally how I fix such a scenario, but that is a bit messy, and I believe this feature will help with this by a lot

**Editor Alternatives (Perhaps?)**
If anyone knows of an visual studio extension (perhaps?) that would do something similar I would be very interested in that.

**No potential .Net Issues (As far as my understanding goes)**
As far as my understanding goes, this should be possible even while following the .net specifications because when partial types get compiled they all get added into the same type as if they were never partial to begin with thus the behaviour of existing runtime accessibility checks will remain the same.
This is thus syntax sugar which won't be embedded anywhere once the code is compiled

**Ending**
This is my first time submitting a proposal, if I'm missing any information of if I have misunderstood the guidelines please let me know so I can correct it as far as possible.
Please feel free to comment what you think about this proposal, I'm interested to see what you people out there think of it

**Comment by TactiTac0z** [comment]
Please excuse my cringe name, this account was made for many years ago, I rarely use GitHub

**Comment by HaloFour** [comment]
See: #5529

**Reply by TactiTac0z** [reply]
Hey my proposal doesn't really seem to be that semilare.
I don't see any sections regarding visability across partial-blocks.
If I'm mistaking feel free to point me to where it is

**Reply by PathogenDavid** [reply]
It's not exactly the same, but typically partial blocks are spread across multiple files -- so in that situation it'd be effectively the same feature.

**Reply by TactiTac0z** [reply]
> It's not exactly the same, but typically partial blocks are spread across multiple files -- so in that situation it'd be effectively the same feature.

I skimmed the proposal linked, as far as I can see it's only really about allowing types to be visible only within the current file. Unlike my proposal which allows both types and members. Members being a big part of what I would use my proposal for.
Besides I use partial within the same file often, so even then that proposal wouldn't really apply to me that much.
You would not be able to have file only fields (or other members) on that proposal (again unless I'm missing something)

**Reply by HaloFour** [reply]
> I skimmed the proposal linked, as far as I can see it's only really about allowing types to be visible only within the current file. Unlike my proposal which allows both types and members. Members being a big part of what I would use my proposal for.

This seems like it would be a natural progression of that proposal.

> Besides I use partial within the same file often, so even then that proposal wouldn't really apply to me that much.
You would not be able to have file only fields (or other members) on that proposal (again unless I'm missing something)

This feels like a very uncommon use of `partial`, you'll probably have to demonstrate why there is value in doing something like this to the extent that it's worth a language feature.  You can prevent that cross-partial access already via analyzers which would likely be the suggested route.

**Reply by TactiTac0z** [reply]
Hey @HaloFour  , do you have any experience in writing custom analysers?
At the end of the day, all I need is members not being recommended in visual studio's suggestion feature (where it suggests type, members names etc when you are typing) under the conditions explain in the proposal (a compiler warning or error would be a bonus - I suppose when assigned a specific attribute since I probably can't easily add the custom blockonly keyword I assume)

I have been trying to look into it but it sure seems pretty complicated.

**Reply by TactiTac0z** [reply]
Assuming of course an analyser for this doesn't already exist which I assume there isn't

**Comment by HaloFour** [comment]
Also, proposals should be opened as [discussions](https://github.com/dotnet/csharplang/discussions).

**Reply by TactiTac0z** [reply]
Ohh I see, thank you.
